### ISSUE TITLE
Add tests for windows signal handling in process

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3186]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#3186|
|Type|enhancement|
|Created|2008-04-16 20:20:25Z|
|Branch|https://github.com/twisted/twisted/tree/win32-signal-tests-3186|

This should work for KILL/INT/TERM, but some tests will ensure it.

<details><summary>Searchable metadata</summary>

```
trac-id__3186 3186
type__enhancement enhancement
reporter__therve therve
priority__normal normal
milestone__ 
branch__branches_win32_signal_tests_3186 branches/win32-signal-tests-3186
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1208377225000000 1208377225000000
changetime__1209025502000000 1209025502000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

(In [23317]) Branching to 'win32-signal-tests-3186'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [23318]) Add tests.

Refs [#3186](https://github.com/twisted/twisted/issues/3186)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @therve|
|-|-|

  1. It would be excellent if the signal test methods (I don't care about the support/helper functions) could be shared between the POSIX and Win32 tests so it's more clear that the same behavior is required/provided on each platform.  Can you factor these into a mixin?
  1. The `failIf` usage in the tests is weird.  I guess changing that to just be Failure-based (ie, have the process protocol errback its deferred - or better yet, have it provide all information about how the process ran and exited and let the decide decide what a failure is) is out of scope for this ticket, though.
  1. It occurs to me that a better thing for `process_signal.py` to do than `time.sleep(5)` would be to try reading from stdin.  It'll block, so the process stays alive long enough for the parent to send it a signal, but if the test fails for some reason, the child will at least exit when the parent exits, since its stdin will disappear and the read will return.  I guess that's out of scope too.

If you address 1 then I think the branch will be in good shape to merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve removed owner|
|-|-|

I did 2 and 3, and added documentation to explain why I think 1 isn't really doable easily and/or cleanly. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @therve|
|-|-|

  * `SignalProtocol`'s docstring would be a bit clearer if it ended with _signal when data is first received._.
  * `SignalProtocol.processEnded` is still pretty complex, so some docs on what it is intended to do (doesn't have to be exactly precise in every detail - but at least describe when it errbacks and when it callbacks) would be nice
  * in the docstring for `test_signalHUP`, drop _the_ in _... the C{processEnded} is called... _
  * similarly for `test_signalINT`, `test_signalKILL`, and `test_signalTERM`.
  * docstring for `Win32SignalProtocol.processEnded` would be nice, just something short explaining errback/callback cases.
  * same grammar comment about the new signal test methods on `Win32ProcessTestCase`

A couple observations about `process_signal.py` (there's not necessarily anything to do about these, they're not really related to this change, but we may want to make tickets for them - mostly I just want to point them out):

  * The default SIGHUP handler is... `SIG_DFL`.  I wonder why this program bothers to explicitly set it to that.
  * A number of the tests run this program with a pipe (not a pty) for stdout, which is probably going to be buffered (not line buffered).  The tests wait to see output before sending the signal, and the program doesn't flush its output.  I wonder if this has caused any intermittent errors.  I know I've seen some tests fail due to trying to kill a process with a `None` pid, which often means the process already exited (since the transport's pid is set to `None` when that happens).  I can't remember which, though.

Otherwise, great.  Please merge when the first group of things is fixed.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set status to `closed`|
|-|-|

(In [23413]) Merge win32-signal-tests-3186

Author: therve
Reviewer: exarkun
Fixes [#3186](https://github.com/twisted/twisted/issues/3186)

Add tests for _dumbwin32proc support of terminate signals, and clean POSIX
signal tests.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
